src/Network/Wai/SAML2/Response.hs

```diff
@@ -62,7 +65,12 @@ instance FromXML Response where
             Nothing -> fail "Invalid status code"
             Just sc -> pure sc
 
-        encAssertion <- oneOrFail "EncryptedAssertion is required"
+        let assertion = oneOrFail "Assertion is required"
+                    (   cursor
+                    $/  element (saml2Name "Assertion")
+                    ) >>= parseXML
+
+        let encAssertion = oneOrFail "EncryptedAssertion is required"
```

Comment:
Just an observation more so than a comment, but the error messages here will never be used. I was briefly wondering if `listToMaybe` might be better, but that has different semantics (at least one element) than `oneOrFail` (exactly one element).

Reply:
It would make no significant difference in practice, but I changed the logic so that it fails when there are more than one `EncryptedAssertion`s

Reply:
I think that's a good change, although my comment applied to both `oneOrFail`s in the snippet above (for `assertion` and `encAssertion`)! Maybe we could have a function `zeroOrOne :: MonadFail m => [a] -> m (Maybe a)` in `Network.Wai.SAML2.XML`?

Reply:
Ah, you're right.